components/org.wso2.carbon.identity.query.saml/src/main/java/org/wso2/carbon/identity/query/saml/handler/SAMLAssertionFinderImpl.java

```diff
@@ -0,0 +1,164 @@
+import org.wso2.carbon.identity.core.persistence.JDBCPersistenceManager;
+import org.wso2.carbon.identity.query.saml.util.SAMLQueryRequestUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation class to find assertions dynamically from assertion stores
+ *
+ * @see SAMLAssertionFinder
+ */
+public class SAMLAssertionFinderImpl implements SAMLAssertionFinder {
+    protected static final Log log = LogFactory.getLog(SAMLAssertionFinderImpl.class);
+
+    /**
+     * This method is used to initialize handler
+     */
+    public void init() {
+
+    }
+
+    /**
+     * Method to finf assertions from assertion stores by assertion is
+     *
+     * @param id This is the unique Assertion ID
+     * @return Assertion This returns assertion with given assertion id
+     */
+    public Assertion findByID(String id) {
+        Assertion assertion = null;
+        if (id.length() > 0 && id != null) {
+            assertion = readAssertion(id);
+            return assertion;
+
+        } else {
+            log.error("Assertion ID field is Empty");
+
+        }
+
+        return assertion;
+    }
+
+    /**
+     * This method is used to search messages from assertion stores by subject of the assertion
+     *
+     * @param subject This is full qualified subject of the request
+     * @return List This returns a list of assertions
+     */
+    public List<Assertion> findBySubject(String subject) {
+        List<Assertion> assertions = new ArrayList<Assertion>();
+        Connection connection = null;
+        PreparedStatement preparedStatement = null;
+        ResultSet resultSet = null;
+        try {
+            connection = JDBCPersistenceManager.getInstance().getDBConnection();
+            preparedStatement = connection.prepareStatement("SELECT SAML2_ASSERTION FROM IDN_SAML2_ASSERTION_STORE WHERE SAML2_SUBJECT= ?");
+            preparedStatement.setString(1, subject);
+            resultSet = preparedStatement.executeQuery();
+            int index = 1;
+            while (resultSet.next()) {
+                String assertionString = resultSet.getString(index);
+                try {
+                    assertions.add((Assertion) SAMLQueryRequestUtil.unmarshall(assertionString));
+                    log.info("Assertions retrieved from database");
+                } catch (Exception e) {
+                    log.error("Unable to Retrieve Assertions from ResultSet ", e);
+                }
+            }
+            if (assertions.size() > 0)
+                return assertions;
+            return null;
+        } catch (SQLException e) {
+            log.error("Error while reading data", e);
+        } finally {
+            try {
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+                if (connection != null) {
```

Comment:
use util method in IdentityDatabaseUtil  to close statement
IdentityDatabaseUtil.closeConnection(connection)